beacon-chain/blockchain/service.go

```diff
@@ -131,83 +141,46 @@ func (c *ChainService) updateHead(processedBlock <-chan *types.Block) {
 		case <-c.ctx.Done():
 			return
 		case block := <-processedBlock:
-			log.WithField("slot", block.SlotNumber()).Info("New beacon slot")
-
 			h, err := block.Hash()
 			if err != nil {
 				log.Errorf("Could not hash incoming block: %v", err)
 				continue
 			}
 
-			log.Info("Applying fork choice rule")
-
-			parentBlock, err := c.beaconDB.GetBlock(block.ParentHash())
-			if err != nil {
-				log.Errorf("Failed to get parent of block %#x", h)
-				continue
-			}
-
-			cState := c.beaconDB.GetCrystallizedState()
-			aState := c.beaconDB.GetActiveState()
-			var stateTransitioned bool
-
-			for cState.IsCycleTransition(block.SlotNumber()) {
-				log.Infof("Recalculating crystallized state")
-				cState, err = cState.NewStateRecalculations(
-					aState,
-					block,
-					c.enableCrossLinks,
-					c.enableRewardChecking,
-				)
-				if err != nil {
-					log.Errorf("Initialize new cycle transition failed: %v", err)
-					continue
-				}
-				stateTransitioned = true
-			}
-
-			aState, err = aState.CalculateNewActiveState(
-				block,
-				cState,
-				parentBlock.SlotNumber(),
-				c.enableAttestationValidity,
-			)
+			log.Info("Updating chain head...")
+			currentHead, err := c.beaconDB.GetCanonicalBlock()
 			if err != nil {
-				log.Errorf("Compute active state failed: %v", err)
+				log.Errorf("Could not retrieve chain head block: %v", err)
 				continue
 			}
 
-			if err := c.beaconDB.SaveActiveState(aState); err != nil {
-				log.Errorf("Write active state to disk failed: %v", err)
+			// If the block's score is less than the current head, we continue and do not
+			// update the chain head.
+			if block.Score() < currentHead.Score() {
```

Comment:
... continuing my earlier comment, I would alter this here. Instead of calculating a score for each block, we can change this if condition to the following:

1. If `block.CrystallizedStateRoot() == currentHead.CrystallizedStateRoot()`, pick the block with the higher slot.
2. Otherwise, fetch each block's crystallized state and:
  2a. Pick the block with the higher `last_finalized_slot`
  2b. If same, pick the block with the higher `last_justified_slot`